File: app/src/test/java/com/mercandalli/android/browser/common/HashMapSharedPreferences.java
package com.mercandalli.android.browser.common;

import android.content.SharedPreferences;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import androidx.annotation.Nullable;

/**
 * A {@link SharedPreferences} that uses an internal {@link Map}
 * to store the items.
 */
public class HashMapSharedPreferences implements SharedPreferences {

    private final Map<String, Object> map;

    public HashMapSharedPreferences() {
        map = new HashMap<>();
    }

    public HashMapSharedPreferences(Map<String, Object> initialMap) {
        map = new HashMap<>();
        if (initialMap != null) {
            map.putAll(initialMap);
        }
    }

    @Override
    public Map<String, ?> getAll() {
        return map;
    }

    @Nullable
    @Override
    public String getString(String key, String defValue) {
        return get(map, key, defValue);
    }

    @Nullable
    @Override
    public Set<String> getStringSet(String key, Set<String> defValues) {
        return get(map, key, defValues);
    }

    @Override
    public int getInt(String key, int defValue) {
        return get(map, key, defValue);
    }

    @Override
    public long getLong(String key, long defValue) {
        return get(map, key, defValue);
    }

    @Override
    public float getFloat(String key, float defValue) {
        return get(map, key, defValue);
    }

    @Override
    public boolean getBoolean(String key, boolean defValue) {
        return get(map, key, defValue);
    }

    @Override
    public boolean contains(String key) {
        return map.containsKey(key);
    }

    @Override
    public MapEditor edit() {
        return new MapEditor(map);
    }

    @Override
    public void registerOnSharedPreferenceChangeListener(OnSharedPreferenceChangeListener listener) {

    }

    @Override
    public void unregisterOnSharedPreferenceChangeListener(OnSharedPreferenceChangeListener listener) {

    }

    private static class MapEditor implements Editor {

        private final Map<String, Object> map;

        private MapEditor(Map<String, Object> map) {
            this.map = map;
        }

        @Override
        public Editor putString(String key, String value) {
            map.put(key, value);
            return this;
        }

        @Override
        public Editor putStringSet(String key, Set<String> values) {
            map.put(key, values);
            return this;
        }

        @Override
        public Editor putInt(String key, int value) {
            map.put(key, value);
            return this;
        }

        @Override
        public Editor putLong(String key, long value) {
            map.put(key, value);
            return this;
        }

        @Override
        public Editor putFloat(String key, float value) {
            map.put(key, value);
            return this;
        }

        @Override
        public Editor putBoolean(String key, boolean value) {
            map.put(key, value);
            return this;
        }

        @Override
        public Editor remove(String key) {
            map.remove(key);
            return this;
        }

        @Override
        public Editor clear() {
            map.clear();
            return this;
        }

        @Override
        public boolean commit() {
            return true;
        }

        @Override
        public void apply() {

        }
    }

    private static <T> T get(Map<String, ?> map, String key, T defValue) {
        if (map.containsKey(key)) {
            return (T) map.get(key);
        } else {
            return defValue;
        }
    }
}

File: app/src/test/java/com/mercandalli/android/browser/common/HashMapSharedPreferences.kt
package com.mercandalli.android.browser.common

import android.content.SharedPreferences

import java.util.HashMap

/**
 * A [SharedPreferences] that uses an internal [Map]
 * to store the items.
 */
class HashMapSharedPreferences : SharedPreferences {

    private val map: MutableMap<String, Any>

    init {
        map = HashMap()
    }

    override fun getAll(): Map<String, *> {
        return map
    }

    override fun getString(key: String, defValue: String?): String? {
        return get(map, key, defValue)
    }

    override fun getStringSet(key: String, defValues: Set<String>?): Set<String>? {
        return get(map, key, defValues)
    }

    override fun getInt(key: String, defValue: Int): Int {
        return get(map, key, defValue)!!
    }

    override fun getLong(key: String, defValue: Long): Long {
        return get(map, key, defValue)!!
    }

    override fun getFloat(key: String, defValue: Float): Float {
        return get(map, key, defValue)!!
    }

    override fun getBoolean(key: String, defValue: Boolean): Boolean {
        return get(map, key, defValue)!!
    }

    override fun contains(key: String): Boolean {
        return map.containsKey(key)
    }

    override fun edit(): SharedPreferences.Editor {
        return MapEditor(map)
    }

    override fun registerOnSharedPreferenceChangeListener(listener: SharedPreferences.OnSharedPreferenceChangeListener) {

    }

    override fun unregisterOnSharedPreferenceChangeListener(listener: SharedPreferences.OnSharedPreferenceChangeListener) {

    }

    private class MapEditor constructor(
            private val map: MutableMap<String, Any>
    ) : SharedPreferences.Editor {

        override fun putString(key: String, value: String?): SharedPreferences.Editor {
            map[key] = value!!
            return this
        }

        override fun putStringSet(key: String, values: Set<String>?): SharedPreferences.Editor {
            map[key] = values!!
            return this
        }

        override fun putInt(key: String, value: Int): SharedPreferences.Editor {
            map[key] = value
            return this
        }

        override fun putLong(key: String, value: Long): SharedPreferences.Editor {
            map[key] = value
            return this
        }

        override fun putFloat(key: String, value: Float): SharedPreferences.Editor {
            map[key] = value
            return this
        }

        override fun putBoolean(key: String, value: Boolean): SharedPreferences.Editor {
            map[key] = value
            return this
        }

        override fun remove(key: String): SharedPreferences.Editor {
            map.remove(key)
            return this
        }

        override fun clear(): SharedPreferences.Editor {
            map.clear()
            return this
        }

        override fun commit(): Boolean {
            return true
        }

        override fun apply() {

        }
    }

    @Suppress("UNCHECKED_CAST")
    private operator fun <T> get(map: Map<String, *>, key: String, defValue: T?): T? {
        return if (map.containsKey(key)) {
            map[key] as T?
        } else {
            defValue
        }
    }
}
